webhook/server/server.go

```diff
@@ -0,0 +1,437 @@
+	}
+
+	/* read AdmissionReview from the request body */
+	ar, err := deserializeAdmissionReview(body)
+	if err != nil {
+		err := logging.Errorf("error deserializing AdmissionReview: %s", err.Error())
+		return nil, http.StatusBadRequest, err
+	}
+
+	return ar, http.StatusOK, nil
+}
+
+func deserializeAdmissionReview(body []byte) (*v1beta1.AdmissionReview, error) {
+	ar := &v1beta1.AdmissionReview{}
+	runtimeScheme := runtime.NewScheme()
+	codecs := serializer.NewCodecFactory(runtimeScheme)
+	deserializer := codecs.UniversalDeserializer()
+	_, _, err := deserializer.Decode(body, nil, ar)
+
+	/* Decode() won't return an error if the data wasn't actual AdmissionReview */
+	if err == nil && ar.TypeMeta.Kind != "AdmissionReview" {
+		err = fmt.Errorf("received object is not an AdmissionReview")
+	}
+
+	return ar, err
+}
+
+func deserializeNetworkAttachmentDefinition(ar *v1beta1.AdmissionReview) (types.NetworkAttachmentDefinition, error) {
+	/* unmarshal NetworkAttachmentDefinition from AdmissionReview request */
+	netAttachDef := types.NetworkAttachmentDefinition{}
+	err := json.Unmarshal(ar.Request.Object.Raw, &netAttachDef)
+	return netAttachDef, err
+}
+
+func deserializePod(ar *v1beta1.AdmissionReview) (v1.Pod, error) {
+	/* unmarshal Pod from AdmissionReview request */
+	pod := v1.Pod{}
+	err := json.Unmarshal(ar.Request.Object.Raw, &pod)
+	return pod, err
+}
+
+func handleValidationError(w http.ResponseWriter, ar *v1beta1.AdmissionReview, orgErr error) {
+	err := prepareAdmissionReviewResponse(false, orgErr.Error(), ar)
+	if err != nil {
+		err := logging.Errorf("error preparing AdmissionResponse: %s", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	writeResponse(w, ar)
+}
+
+func writeResponse(w http.ResponseWriter, ar *v1beta1.AdmissionReview) {
+	logging.Debugf("sending response to the Kubernetes API server")
+	resp, _ := json.Marshal(ar)
+	w.Write(resp)
+}
+
+func validateHandler(w http.ResponseWriter, req *http.Request) {
+	/* read AdmissionReview from the HTTP request */
+	ar, httpStatus, err := readAdmissionReview(req)
+	if err != nil {
+		http.Error(w, err.Error(), httpStatus)
+		return
+	}
+
+	netAttachDef, err := deserializeNetworkAttachmentDefinition(ar)
+	if err != nil {
+		handleValidationError(w, ar, err)
+		return
+	}
+
+	/* perform actual object validation */
+	allowed, err := validateNetworkAttachmentDefinition(netAttachDef)
+	if err != nil {
+		handleValidationError(w, ar, err)
+		return
+	}
+
+	/* perpare response and send it back to the API server */
+	err = prepareAdmissionReviewResponse(allowed, "", ar)
+	if err != nil {
+		logging.Errorf(err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	writeResponse(w, ar)
+}
+
+func getNetworkAttachmentDefinition(namespace, name string) (*types.NetworkAttachmentDefinition, error) {
+	path := fmt.Sprintf("/apis/k8s.cni.cncf.io/v1/namespaces/%s/network-attachment-definitions/%s", namespace, name)
```

Comment:
At some point, we should adopt the official network CRD client to fetch objects: https://github.com/K8sNetworkPlumbingWG/network-attachment-definition-client recently adopted by the networking WG (though it's not exactly ready as of now, there are 4 open PRs to merge before adopting the library).

Reply:
I'll be happy to use it once ready